aquadoggo/src/db/stores/test_utils.rs

```diff
@@ -168,23 +173,97 @@ pub async fn insert_entry_operation_and_view(
     (document_id, document_view_id)
 }
 
-/// Container for `SqlStore` with access to the document ids and key_pairs used in the
-/// pre-populated database for testing.
-pub struct TestSqlStore {
-    pub store: SqlStorage,
-    pub key_pairs: Vec<KeyPair>,
-    pub documents: Vec<DocumentId>,
+#[async_trait::async_trait]
+pub trait AsyncTestFn {
+    async fn call(self, db: TestDatabase);
 }
 
-/// Fixture for constructing a storage provider instance backed by a pre-polpulated database. Passed
-/// parameters define what the db should contain. The first entry in each log contains a valid CREATE
-/// operation following entries contain duplicate UPDATE operations. If the with_delete flag is set
-/// to true the last entry in all logs contain be a DELETE operation.
+#[async_trait::async_trait]
+impl<FN, F> AsyncTestFn for FN
+where
+    FN: FnOnce(TestDatabase) -> F + Sync + Send,
+    F: Future<Output = ()> + Send,
+{
+    async fn call(self, db: TestDatabase) {
+        self(db).await
+    }
+}
+
+pub struct TestDatabaseRunner {
+    /// Number of entries per log/document.
+    no_of_entries: usize,
+
+    /// Number of authors, each with a log populated as defined above.
+    no_of_authors: usize,
+
+    /// A boolean flag for wether all logs should contain a delete operation.
+    with_delete: bool,
+
+    /// The schema used for all operations in the db.
+    schema: SchemaId,
+
+    /// The fields used for every CREATE operation.
+    create_operation_fields: Vec<(&'static str, OperationValue)>,
+
+    /// The fields used for every UPDATE operation.
+    update_operation_fields: Vec<(&'static str, OperationValue)>,
+}
+
+impl TestDatabaseRunner {
+    pub fn with_db_teardown<F: AsyncTestFn + Send + Sync + 'static>(&self, test: F) {
```

Comment:
A docstring here could be friendly for when you encounter this function in some test.